plugins/out_s3/s3.c

```diff
@@ -2097,6 +2118,91 @@ static void flush_init(void *out_context)
     }
 }
 
+static flb_sds_t s3_format(struct flb_s3 *ctx,
+                           int total_records,
+                           const char *tag, int tag_len,
+                           const void *data, size_t bytes,
+                           int csv_add_columns)
+{
+    int ret;
+    int root_type;
+    char *mp_buf;
+    size_t mp_size;
+    flb_sds_t tmp = NULL;
+    flb_sds_t chunk = NULL;
+
+    /* Process chunk */
+    if (ctx->log_key) {
+        tmp = flb_pack_msgpack_extract_log_key(ctx, total_records, data, bytes);
+        if (!tmp) {
+            flb_plg_error(ctx->ins, "Could not generate records by using 'log_key' content");
+            return NULL;
+        }
+
+        if (ctx->format == S3_FORMAT_CSV) {
+            /*
+             * Since log_key option was used, the incoming data is NOT a msgpack object,
+             * likely a JSON string message that we need to convert back so we can do the CSV formatting.
+             */
+
+            ret = flb_pack_json(data, bytes, &mp_buf, &mp_size, &root_type);
+            if (ret < 0) {
+                return NULL;
+            }
```

Comment:
As I commented before, how does this work? With log key, the value of the log_key is used instead of the whole record. But that's for each record. So the buffer as a whole would be a newline delimited list of the value of each log_key. 

So it might be:
```
val1
val2
val3
val4
```

Or it could be a series of JSONs:

```
{ "embeddedjson": "some value which was inside the log key1"}
{ "embeddedjson": "some value which was inside the log key2"
{ "embeddedjson": "some value which was inside the log key3"
```

In either case, you can not treat the buffer as if it was just a single JSON.

Reply:
I have also been thinking about the same problem. Several ideas I have had are:

- Add support for using parsers on the `log_key`.
- Add a new property to specify a format to parse the `log_key` with.
- Throw an error or warning when using CSV output format without any specified way to parse the `log_key`.

Reply:
Any thoughts, @edsiper or @PettitWesley?

Reply:
I think that to move forward with this CSV feature quickly we should just have it not apply to the log_key option. 

As a longer term enhancement, adding support for a parser for the log_key sounds like an interesting idea. How would that work? Use CSV parser to individually convert each log_key value to a CSV?